### ISSUE TITLE
Can I develop app as Kiosk application using Graph calling API?

### ISSUE THREAD

**Issue by cloudsung97** [microsoftgraph/microsoft-graph-comms-samples#53]
I have WPF program as a Kiosk application using Lync sdk to make 2-way video call. I want to make similar application but using Microsoft team. The step is as below:

  -  kiosk app calls recipient -> recipient accepts call and sees video streaming from a person standing in front of kiosk app.
  -  kiosk also gets video streaming sent back from recipient and shows that on app interface.
  -  users at kiosk side will only see custom interface we develop, not Lync/Microsoft Teams UI.

As far as I understand for Teams, native app (kiosk) needs to call bot api and bot will call MS Graph via REST and Graph will create a call to recipient. I don't quite understand how to proceed after that:

  1.  How caller (kiosk) can receive video streaming from recipient?

   2. How caller (kiosk) side turn on camera programmatically?
     - In Lync, we will install Lync libraries and Lync client on application side, then Lync will handle this itself, so the app using Lync will create a call directly to callee.
     - In Teams, it's different as the app needs to call Graph to create a call. So I am really confused how to do this.

**Comment by ksikorsk** [comment]
Hi,

Have you reviewed the AudioVideoPlaybackBot or the HueBot?

The HueBot receives, processes, and sends video.  It also receives audio.
The AudioVideoPlaybackBot streams audio and video from a file source to the meeting.

One note is that in the Teams roster, the kiosk will show up as a bot, not as a user... but it still has audio/video streaming capabilities.

**Comment by cloudsung97** [comment]
Hi, @ksikorsk Thank you for your information. I am reviewing HueBot as you suggested but don't quite get my head around. From you comment,
> One note is that in the Teams roster, the kiosk will show up as a bot, not as a user... but it still has audio/video streaming capabilities.

Do you mean that the kiosk machine installing my native app will need to be webserver to host bot as well? Sorry, it seems to be a stupid question but I am really confused at this point.

Because at first I thought bot hosting and kiosk app will be on different machine, we would need to host bot on just one place let say in azure app service www.bot1.com/api, then kiosk A, kiosk B, kiosk C will call this api with different payload etc., but this way I don't know how kiosk can turn on their own camera.

**Comment by ksikorsk** [comment]
> Do you mean that the kiosk machine installing my native app will need to be webserver to host bot as well? Sorry, it seems to be a stupid question but I am really confused at this point.

It can either be a web server, or it needs to be connected to a web server.  We do have some clients that integrate our service with some hardware. However, they have introduced a gateway service between the client and Graph Calling APIs. Something like this:

`Hardware <----> Gateway (Win x64) <----> Graph`

There are a couple reasons why it has to be done this way:
- App Hosted Media applications are only supported from within a Windows environment.
- Hardware usually does not host an http service, so needs another mechanism to receive notifications from our graph service.

It is your responsibility to ensure that media and signalling flow between hardware and gateway. For a sample Gateway implementation please look at the HueBot or AudioVideoPlaybackBot samples. Both samples process raw video.

**Comment by cloudsung97** [comment]
Thank you again. Well, then could it be possible or reasonable if I take all related code in bot example and put it through into my kioks WPF app instead? so this way my app doesn't need to call bot first but can call Graph api directly.

**Comment by ksikorsk** [comment]
As long as your kiosk WPF app can host a web server and it is running a windows x64 machine it should be ok.

**Comment by cloudsung97** [comment]
@ksikorsk Thank you very much for all detail. much appreciated.

**Comment by cloudsung97** [comment]
@ksikorsk Hi again, I am trying to understand the flow from Huebot sample. So I know we will use ngrok for test and debugging. How about production deployment in case kiosks and bot will be on different machines like you mentioned before ,

> Hardware <----> Gateway (Win x64) <----> Graph

Kiosks app will probably be installed on Win x86 hardware. What do I need to configure this machine to make it accessible from gateway bot?

What I am thinking is that:
- use dynamic dns service, set firewall rule. etc.
- then add CNAME record to bot DNS eg. 0.bot.com = 0.aboveDynamicDomain.com:Port

I am not sure that's the way to go. Thanks.

**Comment by ksikorsk** [comment]
Hey @cloudsung97,

I described a possible solution, but be aware that we don't officially support our APIs hosted on a client.  For that you would require a client SDK, which is not yet available.  At this point I have no timeline for such an SDK.  While the solution I described is possible, the challenges to make sure your device is sending and receiving signaling and media to the gateway are yours to figure out.

Thanks,
Kamil
